protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs

```
@@ -32,78 +32,76 @@ impl Executor {
-										// If the verification failed, return the error status.
-										let ms = MempoolStatus::new(MempoolStatusCode::VmError);
-										(ms, Some(vm_status))
-									} else {
-										// add to the mempool
-										let mut core_mempool = self.core_mempool.write().await;
-
-										debug!("Adding transaction to mempool: {:?} {:?}", transaction, transaction.sequence_number());
-										let status = core_mempool.add_txn(
-											transaction.clone(),
-											0,
-											transaction.sequence_number(),
-											TimelineState::NonQualified,
-											true
-										);
-
-										match status.code {
-											MempoolStatusCode::Accepted => {
-												debug!("Transaction accepted: {:?}", transaction);
-											},
-											_ => {
-												debug!("Transaction not accepted: {:?}", status);
-												Err(TransactionPipeError::TransactionNotAccepted(status))?;
-											}
-										}
-
-										// send along to the receiver
-										transaction_channel.send(transaction).await.map_err(
-											|e| anyhow::anyhow!("Error sending transaction: {:?}", e)
-										)?;
-
-										// report status
-										let ms = MempoolStatus::new(MempoolStatusCode::Accepted);
-										(ms, None)
-									};
-
-									if callback.send(Ok(status)).is_err() {
-										debug!("submit_transaction request has been canceled");
-									}
-								},
-								MempoolClientRequest::GetTransactionByHash(hash, sender) => {
-									let mempool = self.core_mempool.read().await;
-									let mempool_result = mempool.get_by_hash(hash);
-									if sender.send(mempool_result).is_err() {
-										debug!("get_transaction_by_hash request has been canceled");
-									}
+		// Drop the receiver RwLock as soon as possible.
+		let next = {
+			let mut mempool_client_receiver = self.mempool_client_receiver.write().await;
+			mempool_client_receiver.next().await
+		};
+		
+		if let Some(request) = next  {
```

When the stream is closed, assuming it's fused, `next` will be `None` and the `tick_transaction_pipe` call with return `Ok(())` indefinitely. Perhaps it's better to return an error in this case like in #244.

---

Done in [9e10091](https://github.com/movementlabsxyz/movement/pull/243/commits/9e10091af6867347029b877c3582f2037e52bdb9).